File: Software/MenuPrototype/ActionRunnerModal.cs
using System;

namespace MenuPrototype
{
	partial class Program
	{
		/////////////////////////////////////////////////////////////
		public class ActionRunnerModal : MenuItem
		{
			protected MenuItem _followModal;
			protected Func<bool> _isComplete;

			public ActionRunnerModal(string prompt, string tag, MenuItem followModal = null) : base(prompt, tag)
			{
				_isComplete = null;
				_followModal = followModal;
			}

			public ActionRunnerModal(string prompt, string tag, Func<bool> isComplete, MenuItem followModal = null) : base(prompt, tag)
			{
				_isComplete = isComplete;
				_followModal = followModal;
			}

			public virtual bool isComplete() { return true; }

			public override void onDisplay(bool modal = false)
			{
				bool complete = _isComplete == null ? isComplete() : _isComplete();
				if (complete)
				{
					if (_followModal != null)
					{
						getMainMenu().activateDialog(_followModal.getTag());
					}
					else
					{
						getMainMenu().closeDialog();
					}
				}
			}
		}
	}
}

File: Software/Arduino code/OpenAstroTracker/lib/menu/controls/ActionRunnerModal.cpp
#include "../../util/debug.hpp"
#include "MainMenu.hpp"
#include "ActionRunnerModal.hpp"

ActionRunnerModal::ActionRunnerModal(String displayname, String tag, MenuItem *followModal) : MenuItem(displayname, tag)
{
    _isComplete = nullptr;
    _followModal = followModal;
}

ActionRunnerModal::ActionRunnerModal(String displayname, String tag, boolFunc isComplete, MenuItem *followModal) : MenuItem(displayname, tag)
{
    _isComplete = isComplete;
    _followModal = followModal;
}

void ActionRunnerModal::onDisplay(bool modal)
{
    bool complete = _isComplete == nullptr ? isComplete() : _isComplete();
    if (complete)
    {
        if (_followModal != nullptr)
        {
            LOGV2(255, F("ARM: Modal action is complete, activating next dialog %s "), _followModal->getTag().c_str());
            getMainMenu()->activateDialog(_followModal->getTag());
        }
        else
        {
            LOGV1(255, F("ARM: Modal action is complete, closing dialog."));
            getMainMenu()->closeDialog();
        }
    }
}
